pkg/controllers/cloud/virtualmachine_controller.go

```diff
@@ -33,95 +38,174 @@ const (
+// the EE will be deleted.
+func (r *VirtualMachineManager) syncExternalEntities(vmMap map[types.NamespacedName]struct{}) error {
+	eeList := &antreav1alpha2.ExternalEntityList{}
+	if err := r.Client.List(context.TODO(), eeList, &client.ListOptions{}); err != nil {
 		return err
 	}
-	vmList := &cloudv1alpha1.VirtualMachineList{}
-	if err := r.Client.List(context.TODO(), vmList, &client.ListOptions{}); err != nil {
+	for _, ee := range eeList.Items {
+		eeLabelKeyName, exists := ee.Labels[config.ExternalEntityLabelKeyName]
+		if !exists {
+			// Ignore EE objects not created by converter module.
+			continue
+		}
+		eeNn := types.NamespacedName{
+			Name:      eeLabelKeyName,
+			Namespace: ee.Namespace,
+		}
+		if _, ok := vmMap[eeNn]; !ok {
+			r.log.Info("Could not find matching VM object, deleting ExternalEntity", "namespacedName", eeNn)
+			// Delete the external entity, since no matching VM found.
+			_ = r.Client.Delete(context.TODO(), &ee)
+			continue
+		}
+	}
+	return nil
+}
+
+// syncExternalNodes validates that each EN has corresponding VM. If it does not exist then
+// the EN will be deleted.
+func (r *VirtualMachineManager) syncExternalNodes(vmMap map[types.NamespacedName]struct{}) error {
+	enList := &antreav1alpha1.ExternalNodeList{}
+	if err := r.Client.List(context.TODO(), enList, &client.ListOptions{}); err != nil {
 		return err
 	}
+	for _, en := range enList.Items {
+		enLabelKeyName, exists := en.Labels[config.ExternalEntityLabelKeyName]
+		if !exists {
+			// Ignore EN objects not created by converter module.
+			continue
+		}
+		eeNn := types.NamespacedName{
+			Name:      enLabelKeyName,
+			Namespace: en.Namespace,
+		}
+		if _, ok := vmMap[eeNn]; !ok {
+			r.log.Info("Could not find matching VM object, deleting ExternalNode", "namespacedName", eeNn)
+			// Delete the external entity, since no matching VM found.
+			_ = r.Client.Delete(context.TODO(), &en)
+			continue
+		}
+	}
+	return nil
+}
 
-	r.pendingSyncCount = len(vmList.Items)
-	if r.pendingSyncCount == 0 {
-		GetControllerSyncStatusInstance().SetControllerSyncStatus(ControllerTypeVM)
+// resetWatcher sets a watcher to watch VM events.
+func (r *VirtualMachineManager) resetWatcher() (err error) {
+	r.vmWatcher, err = r.Inventory.WatchVms(context.TODO(), "", labels.Everything(), fields.Everything())
+	if err != nil {
+		r.log.Error(err, "failed to start vmWatcher")
+		return err
 	}
-	r.Initialized = true
-	r.Log.Info("Init done", "controller", ControllerTypeVM.String())
 	return nil
 }
 
-// updatePendingSyncCountAndStatus decrements the pendingSyncCount and when
-// pendingSyncCount is 0, sets the sync status.
-func (r *VirtualMachineReconciler) updatePendingSyncCountAndStatus() {
-	if r.pendingSyncCount > 0 {
-		r.pendingSyncCount--
-		if r.pendingSyncCount == 0 {
-			GetControllerSyncStatusInstance().SetControllerSyncStatus(ControllerTypeVM)
+// processEvent waits for any VM events and feeds the event to the
```

Comment:
lets put 80 max character line for any comments

Reply:
Will make this as a standard going forward.